SwiftMonkey/MonkeyXCTestPrivate.swift

```diff
@@ -45,22 +52,23 @@ extension Monkey {
     }
 
     /**
-        Add an action that generates a tap, with a possibility for
-        multiple taps with multiple fingers, using the private
-        XCTest API.
-
-        - parameter weight: The relative probability of this
-          event being generated. Can be any value larger than
-          zero. Probabilities will be normalised to the sum
-          of all relative probabilities.
-        - parameter multipleTapProbability: Probability that
-          the tap event will tap multiple times. Between 0 and 1.
-        - parameter multipleTouchProbability: Probability that
-          the tap event will use multiple fingers. Between 0 and 1.
-    */
+     Add an action that generates a tap, with a possibility for
+     multiple taps with multiple fingers, using the private
+     XCTest API.
+
+     - parameter weight: The relative probability of this
+     event being generated. Can be any value larger than
+     zero. Probabilities will be normalised to the sum
+     of all relative probabilities.
+     - parameter multipleTapProbability: Probability that
+     the tap event will tap multiple times. Between 0 and 1.
+     - parameter multipleTouchProbability: Probability that
+     the tap event will use multiple fingers. Between 0 and 1.
+     */
     public func addXCTestTapAction(weight: Double, multipleTapProbability: Double = 0.05,
-    multipleTouchProbability: Double = 0.05) {
+                                   multipleTouchProbability: Double = 0.05) {
         addAction(weight: weight) { [weak self] in
+            guard self?.isInForeground == true else { return }
```

Comment:
This guard statement is repeated few times in the file. Also it won't prevent triggering actions in the background if someone uses `monkey.addDefaultUIAutomationActions()` to setup the tests.

I would suggest to put this code in the `Monkey.addAction(...)` methods because it is a common place for both  `UIAutomation` and `XCTestPrivate` test setup methods. Here's an example of how this could be done:
```swift
// We should definitely `typealias` the `() -> Void` to be something more verbose... like `Action`
func guardActionFromRunningInBackground(_ action: @escaping () -> Void) -> () -> Void {
    return {
        guard self.isInForeground else {
            self.reactivateApplication()
            return
        }
        action()
    }
}

public func addAction(weight: Double, action: @escaping () -> Void) {
    totalWeight += weight
    randomActions.append((accumulatedWeight: totalWeight,
                          action: guardActionFromRunningInBackground(action)))
}

public func addAction(interval: Int, action: @escaping () -> Void) {
    regularActions.append((interval: interval,
                           action: guardActionFromRunningInBackground(action)))
}
```

With this setup every action you add will only run in the foreground. In addition we'll have the _reactivate app if needed_ logic handled here as well.

So, what do you think @DmitryBespalov ?

Reply:
I like it.